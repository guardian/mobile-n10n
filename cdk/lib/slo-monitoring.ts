import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import type { App } from 'aws-cdk-lib';
import { PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { SqsEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
import { Queue } from 'aws-cdk-lib/aws-sqs';
import {
	ParameterDataType,
	ParameterTier,
	StringParameter,
} from 'aws-cdk-lib/aws-ssm';

export class SloMonitoring extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const queue = new Queue(this, 'MessageQueue', {
			queueName: `notifications-slo-monitoring-${props.stage}`,
		});

		// this advertises the url of the queue to the notifications app
		new StringParameter(this, 'SenderQueueSSMParameter', {
			parameterName: `/notifications/${props.stage}/mobile-notifications/notifications.queues.sloMonitoring`,
			description:
				'Queue for SLO Monitoring service. N.B. this parameter is created via cdk',
			simpleName: false,
			stringValue: queue.queueUrl,
			tier: ParameterTier.STANDARD,
			dataType: ParameterDataType.TEXT,
		});

		// this advertises the arn of the queue which can be used by the notifications app cloudformation
		new StringParameter(this, 'SenderQueueSSMArnParameter', {
			parameterName: `/notifications/${props.stage}/mobile-notifications/notifications.queues.sloMonitoringArn`,
			description:
				'Arn for the queue for SLO Monitoring service. N.B. this parameter is created via cdk',
			simpleName: false,
			stringValue: queue.queueArn,
			tier: ParameterTier.STANDARD,
			dataType: ParameterDataType.TEXT,
		});

		const sloMonitor = new GuLambdaFunction(this, 'SloMonitor', {
			functionName: `mobile-notifications-slo-monitor-${props.stage}`,
			fileName: 'slomonitor.jar',
			handler: 'com.gu.notifications.slos.SloMonitor::handleMessage',
			runtime: Runtime.JAVA_11,
			app: 'slomonitor',
		});

		sloMonitor.addToRolePolicy(
			new PolicyStatement({
				actions: ['cloudwatch:PutMetricData'],
				resources: ['*'],
			}),
		);

		queue.grantConsumeMessages(sloMonitor);

		sloMonitor.addEventSource(new SqsEventSource(queue, { batchSize: 1 }));
	}
}
