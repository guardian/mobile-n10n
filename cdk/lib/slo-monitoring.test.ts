import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { SloMonitoring } from './slo-monitoring';

describe('The SloMonitoring stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new SloMonitoring(app, 'SloMonitoring', {
			stack: 'mobile-notifications',
			stage: 'PROD',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
