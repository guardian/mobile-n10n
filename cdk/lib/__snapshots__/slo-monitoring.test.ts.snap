// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SloMonitoring stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DeadLetterQueue9F481546": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "notifications-slo-monitoring-dlq-PROD",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-n10n",
          },
          {
            "Key": "Stack",
            "Value": "mobile-notifications",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "MessageQueue7A3BF959": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "notifications-slo-monitoring-PROD",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "DeadLetterQueue9F481546",
              "Arn",
            ],
          },
          "maxReceiveCount": 1,
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-n10n",
          },
          {
            "Key": "Stack",
            "Value": "mobile-notifications",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "VisibilityTimeout": 360,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "SenderQueueSSMArnParameter4B75A98E": {
      "Properties": {
        "DataType": "text",
        "Description": "Arn for the queue for SLO Monitoring service. N.B. this parameter is created via cdk",
        "Name": "/notifications/PROD/mobile-notifications/notifications.queues.sloMonitoringArn",
        "Tags": {
          "Stack": "mobile-notifications",
          "Stage": "PROD",
          "gu:cdk:version": "TEST",
          "gu:repo": "guardian/mobile-n10n",
        },
        "Tier": "Standard",
        "Type": "String",
        "Value": {
          "Fn::GetAtt": [
            "MessageQueue7A3BF959",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SenderQueueSSMParameterD141981E": {
      "Properties": {
        "DataType": "text",
        "Description": "Queue for SLO Monitoring service. N.B. this parameter is created via cdk",
        "Name": "/notifications/PROD/mobile-notifications/notifications.queues.sloMonitoring",
        "Tags": {
          "Stack": "mobile-notifications",
          "Stage": "PROD",
          "gu:cdk:version": "TEST",
          "gu:repo": "guardian/mobile-n10n",
        },
        "Tier": "Standard",
        "Type": "String",
        "Value": {
          "Ref": "MessageQueue7A3BF959",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SloMonitor1106EE7E": {
      "DependsOn": [
        "SloMonitorServiceRoleDefaultPolicy8A58BB27",
        "SloMonitorServiceRole66A7ADCC",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "mobile-notifications/PROD/slomonitor/slomonitor.jar",
        },
        "Environment": {
          "Variables": {
            "APP": "slomonitor",
            "STACK": "mobile-notifications",
            "STAGE": "PROD",
          },
        },
        "FunctionName": "mobile-notifications-slo-monitor-PROD",
        "Handler": "com.gu.notifications.slos.SloMonitor::handleMessage",
        "LoggingConfig": {
          "LogFormat": "JSON",
        },
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "SloMonitorServiceRole66A7ADCC",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Tags": [
          {
            "Key": "App",
            "Value": "slomonitor",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-n10n",
          },
          {
            "Key": "Stack",
            "Value": "mobile-notifications",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "SloMonitorServiceRole66A7ADCC": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "slomonitor",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-n10n",
          },
          {
            "Key": "Stack",
            "Value": "mobile-notifications",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SloMonitorServiceRoleDefaultPolicy8A58BB27": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/mobile-notifications/PROD/slomonitor/slomonitor.jar",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/mobile-notifications/slomonitor",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/mobile-notifications/slomonitor/*",
                  ],
                ],
              },
            },
            {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "athena:StartQueryExecution",
                "athena:GetQueryExecution",
                "athena:GetQueryResults",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "glue:GetDatabase",
                "glue:GetTable",
                "glue:GetPartitions",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:glue:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":catalog",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:glue:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":database/default",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:glue:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":database/notifications",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:glue:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/notifications/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:Get*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                "arn:aws:s3:::aws-mobile-event-logs-prod",
                "arn:aws:s3:::aws-mobile-event-logs-prod/*",
              ],
            },
            {
              "Action": [
                "s3:AbortMultipartUpload",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::aws-mobile-event-logs-prod/athena/slo-monitoring/*",
            },
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MessageQueue7A3BF959",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SloMonitorServiceRoleDefaultPolicy8A58BB27",
        "Roles": [
          {
            "Ref": "SloMonitorServiceRole66A7ADCC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SloMonitorSqsEventSourceSloMonitoringMessageQueue50CCCDD511949E74": {
      "Properties": {
        "BatchSize": 1,
        "EventSourceArn": {
          "Fn::GetAtt": [
            "MessageQueue7A3BF959",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "SloMonitor1106EE7E",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
  },
}
`;
